src/Components/Components/src/NavigationManager.cs

```diff
@@ -274,6 +282,87 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
         }
     }
 
+    /// <summary>
+    /// Notifies the registered handlers of the current location change.
+    /// </summary>
+    /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI.</param>
+    /// <param name="isNavigationIntercepted">Whether this navigation was intercepted from a link.</param>
+    /// <param name="continueNavigation">An action that continues the navigation based on the provided <see cref="LocationChangingResult"/>.</param>
+    protected void NotifyLocationChanging(string uri, bool isNavigationIntercepted, Action<LocationChangingResult> continueNavigation)
```

Comment:
This signature doesn't look right, how would it look if this was async all the way through? (See OnLocationChanged for reference in ComponentHub)

Reply:
The previous implementation was completely async, but the problem is that on Blazor Server, the SignalR property `MaximumParallelInvocationsPerClient` is set to 1. This effectively makes things like JS interop calls blocked until this method completes (and if you try to do a JS interop call in an "OnLocationChanging" handler, the app will deadlock). This is the reason for making this completely synchronous and having a separate "end location changing" call to let the JS side know about the result.

Reply:
Would it work if it doesn't do this via JS interop but via methods in the hub?

Reply:
I agree it seems like it should be possible to hide the SignalR-serial-calls limitation from `NavigationManager` subclasses. If this method was `async`, then perhaps the hub method could deal with awaiting that task and then calling the completion (but without actually returning the child task back to SignalR, so it doesn't block).